### PR TITLE
[None][feat] Support rejection sampling under attention DP (incl. LM-head TP)

### PR DESCRIPTION
**Pull request by zhaoyangwang-nvidia**
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved speculative decoding consistency across parallel processing, preventing ranks from selecting different sampling paths.
  * Fixed advanced draft sampling when language-model head parallelism and attention data parallelism are enabled.
  * Rejection sampling is now allowed with attention parallelism while remaining restricted with context parallelism.

* **Tests**
  * Added coverage for synchronized sampling decisions and parallelism-related rejection sampling behavior.

## Description

Rejection sampling (introduced in #15775; makes speculative decoding
distribution-preserving under non-greedy sampling) was config-gated off under
attention DP. This PR lifts that gate and makes the advanced (rejection) draft
path work correctly under ADP, including the `enable_lm_head_tp_in_adp` mode.

### Why it was gated

Under ADP + LM-head TP, the MTP draft head produces group-stacked,
vocab-sharded logits (`[lm_head_tp_size * max_num_requests, vocab/g]`). Greedy
drafting digests this layout with a cheap distributed argmax (#16440), but
rejection sampling needs each request's **full-vocab proposal distribution**
q(x) to compute `min(1, p/q)` and the `(p - q)+` residual, which vocab shards
cannot provide. Additionally, the greedy-vs-advanced branch is driven by
`is_all_greedy_sample` — a per-rank batch property under ADP — so a diverged
choice would strand part of the LM-head-TP group in a collective the rest
never enters (hang), or replay CUDA graphs with mismatched collective
sequences.

Note: the parallel gate only ever covered the newly wired methods (vanilla
MTP / PARD / DFlash / DraftTarget one-model); MTP-Eagle was never gated, so on
current main MTP-Eagle + ADP + LM-head-TP + rejection + a non-greedy batch can
already reach the advanced path and silently sample from the stacked/sharded
layout. For MTP-Eagle this PR is a correctness fix; for the other methods it
is new support.

### What changed

1. **Group-synchronized `is_all_greedy_sample`** (`interface.py`,
   `model_engine.py`): under ADP + LM-head-TP + rejection, the model engine
   all-gathers the per-rank flags BEFORE the CUDA graph key is built and
   stores the group AND on `spec_metadata.group_all_greedy_sample`;
   `_scan_one_model_sampling` re-applies it on every rescan so the graph key,
   buffer population, and worker branches all observe the same value. AND
   semantics are conservative-safe: a locally all-greedy rank pulled onto the
   advanced path still samples greedily via its sentinel params. The sync
   condition is pure config, so ranks also agree on whether the exchange
   happens at all.

2. **LM-head-TP bypass for advanced draft sampling** (`eagle3.py`): when
   `wants_advanced_draft_sampling` (rejection enabled AND group-synced batch
   not all-greedy), the worker skips the stacked/sharded fast path and calls
   `draft_model.lm_head` directly — under ADP the LMHead weight is replicated
   (the shard slicing is a runtime optimization behind
   `is_spec_decoding_head=True`), so this is a plain local full-vocab GEMM,
   identical to what the target head runs, with zero extra communication.
   With rejection off, non-greedy batches keep the LM-head-TP argmax path
   unconditionally and the per-rank flag never enters control flow.

3. **Lift the config gate** (`llm_args.py`): remove `enable_attention_dp`
   from the rejection parallel gate (context parallelism remains gated).
   Plain ADP needs no synchronization: each rank holds replicated full-vocab
   draft logits for its own requests and the advanced path is local-only.

The greedy-vs-advanced predicate is centralized in
`SpecMetadata.wants_advanced_draft_sampling` so the sampler branch and the
worker's head selection cannot diverge.

## Test Coverage

Unit (CPU): `tests/unittest/_torch/speculative/test_group_all_greedy_sync.py`
(new, 5 cases: local vs group override, rescan survival, capture-override
composition); `tests/unittest/llmapi/test_llm_args.py` (2 new gate cases: ADP
incl. LM-head-TP accepted, context parallelism still raises);
`test_rejection_buffers_guard.py` passes unchanged.

E2E (4xH200, DeepSeek-V3-Lite bf16, tp4 + attention DP + LM-head TP, MTP
nextn=2, rejection enabled, CUDA graphs on):

| batch shape | tokens/iter | validates |
|---|---|---|
| all-greedy | 2.40 | distributed-argmax fast path unregressed |
| all-sampled (temp=0.8, top_p=0.9) | 2.41 | bypass + rejection kernels; acceptance on par with greedy |
| mixed greedy/sampled | 2.41 | per-rank flag divergence: no hang / desync under group sync |

## Related

- Builds on #16440 (greedy distributed argmax for ADP + LM-head TP, nvbug 6460072)
- Rejection sampling feature: #15775

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Note on `ruff-legacy-baseline.json`

This PR resolves 4 legacy lint violations in the files it touches, so the `ruff-legacy` baseline was tightened per the hook's suggestion (`scripts/legacy_utils.py lint-update-violations`). The regeneration is a full rescan and also drops stale baseline entries for files that no longer exist in the tree — that part is mechanical cleanup with no behavior change.

### CODE REVIEW

File: tensorrt_llm/_torch/speculative/interface.py [hunk 1]
```diff
@@ -511,6 +511,27 @@ class SpecMetadata:
...
+    # group waiting in a collective the rest never enters. The model engine
+    # (``_sync_group_all_greedy_sample``) therefore all-gathers the per-rank
+    # local flags each iteration BEFORE the CUDA graph key is built and stores
+    # the group AND here; ``_scan_one_model_sampling`` re-applies it on every
+    # rescan so all later consumers (graph key, populate, worker branches)
+    # observe the same value. AND semantics are conservative-safe: a locally
+    # all-greedy rank may be pulled onto the advanced path, where its sentinel
+    # sampling params still sample greedily -- never the reverse. ``None``
+    # (default) means "no group sync configured": the local per-rank value is
+    # used as-is.
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/speculative/interface.py]: This comment is very verbose when the behavior is pretty simple and easy to understand - consider simplifying it

**Review by mikeiovine — APPROVED** [review]

**Review by tburt-nv — APPROVED** [review]

File: tensorrt_llm/_torch/speculative/interface.py (hunk 1 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/speculative/interface.py]: Done — trimmed to the invariant, writer/applier pointers, and the AND-safety note.

**Review by Tabrizian — APPROVED** [review]
approving on behalf of disagg devs.
